api/authorizers/build.js

```diff
@@ -1,26 +1,59 @@
-const { User, Site } = require('../models');
-
-const authorize = (user, build) => (
-  User.findById(user.id, { include: [Site] })
-    .then((userModel) => {
-      const buildSiteId = build.site || build.Site.id;
-      const matchingUserSite = userModel.Sites.find(site => buildSiteId === site.id);
-
-      if (matchingUserSite) {
-        return Promise.resolve();
-      }
-      return Promise.reject(403);
-    }
-  )
-);
+const { User, Build } = require('../models');
 
-const findOne = (user, build) => authorize(user, build);
+const shaRegex = /^[a-f0-9]{40}/;
+const validateSha = maybeSha => shaRegex.test(maybeSha);
+
+const verifyBuild = (model) => {
+  const build = model.Builds[0];
+
+  if (!build) {
+    return Promise.reject(403);
+  }
+
+  return build.get({ plain: true });
+};
+
+const getBuildById = (user, params) => {
+  const { buildId, siteId } = params;
+
+  return User.findById(user.id, {
+    include: {
+      model: Build,
+      where: { site: siteId, id: buildId },
+    },
+  })
+  .then(verifyBuild);
+};
+
+const getBuildByBranch = (user, params) => {
+  const { siteId, sha, branch } = params;
 
-const create = (user, params) => {
-  if (user.id !== params.user) {
-    throw 403;
+  if (!validateSha(sha)) {
+    throw 400;
   }
-  return authorize(user, params);
+
+  return User.findById(user.id, {
+    include: {
+      model: Build,
+      where: {
+        branch,
+        site: siteId,
+      },
+      order: User.sequelize.literal('"createdAt" desc'),
```

Comment:
would this work: `order: [['createdAt', 'DESC']],` ?

I'm not that familiar with the ordering API in sequelize, but it looks like that might work based on the first "example" at http://docs.sequelizejs.com/manual/tutorial/querying.html#ordering (`['title', 'DESC'],`)

Reply:
yeah it should work, I just didnt like the tuple syntax, although I think they call it out as their preferred method 😬

Reply:
hehe, I don't feel strongly about either way. I personally like the tuple one better (barely), but it's not a big deal.